always-hydrate.md

```diff
@@ -0,0 +1,133 @@
+```
+
+The only change from today's syntax is the optional `[always-]` on the hydrate flag. 
+`--always-hydrate` implies `--hydrate`.
+
+Passing in the `--always-hydrate` flag has all the same behaviors that `prefetch --hydrate` has today
+and also adds the specified files and folders to the `always-hydrate` file, so that those files will
+continue to be hydrated in the future.
+
+Managing existing entries:
+
+In addition to adding entries to the `always-hydrate` file, users will also want to list out the
+existing entries and remove some or all of the existing entries. To that end, we can support the
+following commands:
+
+```
+    gvfs prefetch
+      [<Repo Root Path>]
+      --list-always-hydrate
+```
+
+The `--list-always-hydrate` flag is incompatible with any of the other flags. It will read out the
+`always-hydrate` file and display the entries to the user.
+
+```
+    gvfs prefetch
+      [<Repo Root Path>]
+      [--files <semicolon-delimited list of file paths, prefix * wildcards allowed>] 
+      [--folders <semicolon-delimited list of folder paths, no wildcards>]
+      --remove-always-hydrate
+```
+
+The `--remove-always-hydrate` flag is incompatible with any combination of `--[always-]hydrate` or
+`--list-always-hydrate`.
+
+Any entries that match the specified `--files` and `--folders` arguments will be removed from the
+always-hydrate file. To remove all entries, the user can specify `gvfs prefetch 
+--remove-always-hydrate --files *`.
+
+## Implementation
+### Storing always-hydrate entries
+We will create a new file in the `.gvfs` folder called `always-hydrate`. This file is technically
+human readable and editable, but like all files in the `.gvfs` folder, carries no contract around
+maintaining the same format moving forward. All user interactions with this file should go through
+the `gvfs prefetch` command.
+
+Therefore the specific format of the file is not that interesting. For the sake of easy
+implementation and maintenance, we will reuse our existing file formats from our family of
+`FileBasedCollection` classes, either `FileBasedDictionary` or a new type that derives from
+`FileBasedCollection` in the pattern of the `PlaceholderListDatabase`.
+
+In this case, we need to store an entry per file or folder that was specified to the prefetch
+command, and we will need to remember if it was a file or a folder. Current plan is to go with a
+`FileBasedDictionary` unless we find the need to store more fields.
+
+### Hydrating files after a projection change
+Whenever the projection changes, the `GitIndexProjection` class currently kicks in and updates all
+placeholder files on disk with their new blob ids (for any files that changed). This is all done
+synchronously and blocks the git command that caused the projection change.
+
+In addition, this code will now also schedule background downloads of all paths that match the
+`always-hydrate` collection. We will just pass the list of file and folder parameters to the existing
```

Comment:
> of all paths that match the `always-hydrate` collection

As mentioned in the code changes, limit this list to the placeholder that needed to be updated?

Reply:
And yea as we discussed, this could end up hydrating files that don't currently exist as placeholders, so we do need to start from the index and apply our filters on that. Luckily we already have it all parsed and in memory, so we should be able to just use that.